sync15-adapter/src/sync_multiple.rs

```diff
@@ -0,0 +1,144 @@
+pub struct ClientInfo {
+    // the client_init used to create the client.
+    client_init: Sync15StorageClientInit,
+    // the client we will reuse if possible.
+    client: Sync15StorageClient,
+}
+
+/// Sync multiple stores
+/// * `stores` - The stores to sync
+/// * `persisted_global_state` - The global state to use, or None if never
+///   before provided. At the end of the sync, the value in this cell should
+///   be persisted to permanent storage and provided next time the sync is
+//    called.
+/// * `last_client_info` - The client state to use, or None if never before
+///   provided. At the end of the sync, the value should be persisted
+///   *in memory only* - it should not be persisted to disk.
+/// * `storage_init` - Information about how the sync http client should be
+///   configured.
+/// * `root_sync_key` - The KeyBundle used for encryption.
+pub fn sync_multiple(
+    stores: &[&dyn Store],
+    persisted_global_state: &Cell<Option<String>>,
+    last_client_info: &Cell<Option<ClientInfo>>,
+    storage_init: &Sync15StorageClientInit,
+    root_sync_key: &KeyBundle
+) -> result::Result<HashMap<String, Error>, Error> {
+
+    // Note: We explicitly swap a None back as the state, meaning if we
+    // unexpectedly fail below, the next sync will redownload meta/global,
+    // crypto/keys, etc. without needing to. Apparently this is both okay
+    // and by design.
+    let persisted = persisted_global_state.replace(None);
+    let maybe_global = match persisted {
+        Some(persisted_string) => {
+            match serde_json::from_str::<GlobalState>(&persisted_string) {
+                Ok(state) => Some(state),
+                _ => {
+                    // Don't log the error since it might contain sensitive
+                    // info like keys (the JSON does, after all).
+                    error!("Failed to parse GlobalState from JSON! Falling back to default");
+                    None
+                }
+            }
+        },
+        None => None
+    };
+
+    let mut global_state = match maybe_global {
+        Some(g) => g,
+        None => {
+            info!("First time through since unlock. Creating default global state.");
+            last_client_info.replace(None);
+            GlobalState::default()
+        }
+    };
+
+    // Ditto for the ClientInfo - if we fail below the GlobalStateProvider will
+    // not have the last client and client_init, so will be re-initialized on
+    // the next sync.
+    let client_info = match last_client_info.replace(None) {
+        Some(client_info) => {
+            if client_info.client_init != *storage_init {
+                ClientInfo {
+                    client_init: storage_init.clone(),
+                    client: Sync15StorageClient::new(storage_init.clone())?,
+                }
+            } else {
+                // we can reuse it.
+                client_info
+            }
+        },
+        None => {
```

Comment:
It almost feels like we can introduce a match guard to remove the duplication here, but it doesn't seem like it can work.